pennylane/devices/qubit/preprocess.py

```diff
@@ -77,6 +78,73 @@ def _operator_decomposition_gen(
 #######################
 
 
+def validate_and_expand_adjoint(
+    circuit: qml.tape.QuantumScript,
+) -> qml.tape.QuantumScript:  # pylint: disable=protected-access
+    """Function for validating that the operations and observables present in the input circuit
+    are valid for adjoint differentiation.
+
+    Args:
+        circuit(.QuantumTape): the tape to validate
+
+    Returns:
+        .QuantumTape: The expanded tape, such that it is supported by adjoint differentiation
+
+    Raises:
+        QuantumFunctionError: if the circuit is invalid for adjoint differentiation
+    """
+    # Check validity of measurements
+    measurements = []
+    for m in circuit.measurements:
+        if not isinstance(m, ExpectationMP):
+            raise qml.QuantumFunctionError(
+                "Adjoint differentiation method does not support "
+                f"measurement {m.__class__.__name__}."
+            )
+
+        if not m.obs.has_matrix:
+            raise qml.QuantumFunctionError(
```

Comment:
```suggestion
            raise DeviceError(
```

Since all the other preprocessing things raise `DeviceError`, we should probably raise device error for these too.